xdsl/parser/attribute_parser.py

```diff
@@ -485,24 +485,66 @@ def parse_shape_dimension(self, allow_dynamic: bool = True) -> int:
 
         return int_token.kind.get_int_value(int_token.span)
 
-    def parse_shape_delimiter(self) -> None:
+    def _parse_optional_shape_delimiter(self) -> str | None:
         """
         Parse 'x', a shape delimiter. Note that if 'x' is followed by other
         characters, it will split the token. For instance, 'x1' will be split
         into 'x' and '1'.
         """
         if self._current_token.kind != MLIRTokenKind.BARE_IDENT:
-            self.raise_error(
-                f"Expected 'x' in shape delimiter, got {self._current_token.kind.name}"
-            )
+            return None
 
         if self._current_token.text[0] != "x":
-            self.raise_error(
-                f"Expected 'x' in shape delimiter, got {self._current_token.text}"
-            )
+            return None
 
         # Move the lexer to the position after 'x'.
         self._resume_from(self._current_token.span.start + 1)
+        return "x"
+
+    def parse_shape_delimiter(self) -> None:
+        """
+        Parse 'x', a shape delimiter. Note that if 'x' is followed by other
+        characters, it will split the token. For instance, 'x1' will be split
+        into 'x' and '1'.
+        """
+        token = self._current_token
+        tk = token.kind
+
+        err_val = tk.name if tk != MLIRTokenKind.BARE_IDENT else token.text
+
+        self.expect(
+            self._parse_optional_shape_delimiter,
+            f"Expected 'x' in shape delimiter, got {err_val}",
+        )
```

Comment:
Let's build the error string only in the failing case, it tends to be annoyingly expensive, unfortunately
```suggestion
		if self._parse_optional_shape_delimiter() is None:
		    return
		
        token = self._current_token
        tk = token.kind

        err_val = tk.name if tk != MLIRTokenKind.BARE_IDENT else token.text

        self.raise(
            f"Expected 'x' in shape delimiter, got {err_val}",
        )
```